torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -825,10 +835,20 @@ std::pair<IterDomain*, IterDomain*> IterDomain::split(
       in->getIterType(),
       in->isRFactorProduct());
 
-  new Split(ido, idi, in, factor, inner_split);
+  new Split(ido, idi, in, factor, inner_split, start_offset, stop_offset);
   return {ido, idi};
 }
 
+std::pair<IterDomain*, IterDomain*> IterDomain::split(
+    IterDomain* in,
+    Val* factor,
+    bool inner_split,
+    bool trim_invalid) {
+  auto start_offset = trim_invalid ? in->start() : nullptr;
+  auto stop_offset = trim_invalid ? in->stopOffset() : nullptr;
+  return IterDomain::split(in, factor, inner_split, start_offset, stop_offset);
```

Comment:
Split needs to have its own `start_offset` and `stop_offset` as they are not necessarily determined by `IterDomain::start_offset` and `IterDomain::stop_offset`. Even when `IterDomain::start_offset` is zero, `Split::start_offset` can be non-zero as long as any value at positions from 0 to `start_offset` is not needed to compute outputs (this is validated by `validatePartialSplit`). So, I think it makes sense for each split expression to have its own start and stop offsets.

Reply:
I didn't see in the code how this could happen. Do you have an example of this I can go through?

Reply:
I use it to replaying, e.g., replay partial splits on non-partial domains. For example:

```
  auto tv0 = makeSymbolicTensor(1);
  fusion.addInput(tv0);

  auto tv1 = add(tv0, new Double(1));
  auto tv2 = shift(tv1, {1}, false);
  fusion.addOutput(tv2);
```

This would result in:

```
tv0: [I0]
tv1: [I0]
tv2: [I0-1]
```

Suppose splitting the tv2 axis as:

```
  tv2->split(0, 4, true, true);
```

The fusion math would be:

```
tv0: [I0]
tv1: [I0]
tv2: [ceilDiv(I0-1, 4), 4]
```

With this fusion, `replayPasC(tv1, tv2)` in `getNonGlobalProducerIndex` temporarily transforms `tv1` to `[ceilDiv(I0-1, 4), 4]`, where the same partial split as `tv2` is used even though the root domain of `tv1` just has zero offsets.

Does this answer your question?

Reply:
And this is the actual code generated from the fusion:

```
__global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 1> T2) {
  NVFUSER_DEFINE_MAGIC_ZERO
  float T1[(T0.size[0] + 1)];
  #pragma unroll 1
  for(nvfuser_index_t ki41 = 0; ki41 < (T0.size[0] + 1); ++ki41) {
    if (((ki41 >= 1) && (ki41 < (T0.size[0] + 1)))) {
      T1[ki41]
        = T0[((ki41 + -1) * T0.stride[0])]
        + (float) 1;
    }
  }
  #pragma unroll 1
  for(nvfuser_index_t ki35 = 0; ki35 < (ceilDiv((T0.size[0] - 1), 4)); ++ki35) {
    #pragma unroll
    for(nvfuser_index_t ki37 = 0; ki37 < 4; ++ki37) {
      if ((((((ki35 * 4) + ki37) + 1) >= 1) && ((((ki35 * 4) + ki37) + 1) < T0.size[0]))) {
        T2[((((ki35 * 4) + (ki37 + nvfuser_zero)) + 1) * 1)]
           = T1[((((ki35 * 4) + ki37) + (1 - 1)) + 1)];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
  }
}
```

Specifically, the producer indexing of `T1`:

```
           = T1[((((ki35 * 4) + ki37) + (1 - 1)) + 1)];
```

The `+ (1 - 1)) + 1` part is:

- `(1 - 1)`: `T1` has halo of size 1 at the left end, so `+1` is needed. Since this is `shift +1`, the producer position is `-1` relative to the consumer, so aggregating them becomes `(1 - 1)`.
- `+1`: That's because `T1` is created without any partial split but used with a partial split that trims one element at the left end. In other words, `T1` as this producer has a relative offset of `+1` to `T1` as the consumer.